### PR TITLE
Add write_input() before transferring files to remote

### PR DESCRIPTION
**Pull request by jan-janssen**
This pull request addresses the issue raised by @Leimeroth in https://github.com/pyiron/pyiron_atomistics/issues/1471 that the remote submission for `LAMMPS` jobs failed after the switch to using the `calculate()` function. While it is not exactly clear why the `VASP` job works fine, the work around in this branch fixes the issue for `LAMMPS`.

### CODE REVIEW

File: pyiron_base/jobs/job/runfunction.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-from datetime import datetime
 import multiprocessing
 import os
 import posixpath
 import shutil
 import subprocess
-from typing import Optional, List, Tuple
+from concurrent.futures import ProcessPoolExecutor
+from datetime import datetime
+from typing import List, Optional, Tuple
```
Comment (jan-janssen) [inline pyiron_base/jobs/job/runfunction.py]: @Leimeroth I assume these changes are just related to committing from an older version and not related to this pull request, correct? Then I would revert those for this pull request.

Comment (Leimeroth) [reply pyiron_base/jobs/job/runfunction.py]: That is related to trying https://docs.astral.sh/ruff/ for linting. Forgot to disable the automatic sorting

Comment (jan-janssen) [reply pyiron_base/jobs/job/runfunction.py]: I adopted the ruff sorting for our continuous integration in https://github.com/pyiron/pyiron_base/pull/1512 so these changes are now already included in the `main` branch.

File: pyiron_base/jobs/job/generic.py [hunk 2]
```diff
@@ -632,10 +633,15 @@ def write_input(self):
         Call routines that generate the code specific input files
         Returns:
         """
-        write_input_files_from_input_dict(
-            input_dict=self.get_input_parameter_dict(),
-            working_directory=self.working_directory,
-        )
+        if not self._input_is_written:
```
Comment (jan-janssen) [inline pyiron_base/jobs/job/generic.py]: @Leimeroth Is there an issue with files being written twice? Meaning the remote job trying to write the files again and in the worst case overwriting the existing input files.

Comment (Leimeroth) [reply pyiron_base/jobs/job/generic.py]: I think the write_input_funct is called again somewhere on the cluster. This call tries to copy the potential file, 
so I get the errors when using my own potentials.

Comment (jan-janssen) [inline pyiron_base/jobs/job/generic.py]: How about checking if the working directory is empty and only write the input files when it is:
```suggestion
        if len(os.listdir(self.working_directory)) != 0:
```

File: pyiron_base/jobs/job/generic.py [hunk 3]
```diff
@@ -168,6 +168,7 @@ def __init__(self, project, job_name):
         self._process = None
         self._compress_by_default = False
         self._job_with_calculate_function = False
+        self._input_is_written = False
```
Comment (jan-janssen) [inline pyiron_base/jobs/job/generic.py]:
```suggestion
```

File: pyiron_base/jobs/job/generic.py [hunk 4]
```diff
@@ -1166,6 +1172,7 @@ def to_dict(self):
             "restart_file_dict": self._restart_file_dict,
             "exclude_nodes_hdf": self._exclude_nodes_hdf,
             "exclude_groups_hdf": self._exclude_groups_hdf,
+            "input_is_written": self._input_is_written,
```
Comment (jan-janssen) [inline pyiron_base/jobs/job/generic.py]:
```suggestion
```

File: pyiron_base/jobs/job/generic.py [hunk 5]
```diff
@@ -1195,6 +1202,7 @@ def from_dict(self, job_dict):
             self._restart_file_dict = generic_dict["restart_file_dict"]
             self._exclude_nodes_hdf = generic_dict["exclude_nodes_hdf"]
             self._exclude_groups_hdf = generic_dict["exclude_groups_hdf"]
+            self._input_is_written = generic_dict["input_is_written"]
```
Comment (jan-janssen) [inline pyiron_base/jobs/job/generic.py]:
```suggestion
```

File: pyiron_base/jobs/job/generic.py [hunk 6]
```diff
@@ -632,10 +633,15 @@ def write_input(self):
...
-            working_directory=self.working_directory,
-        )
+        if not self._input_is_written:
+            write_input_files_from_input_dict(
+                input_dict=self.get_input_parameter_dict(),
+                working_directory=self.working_directory,
+            )
+        else:
+            self._logger.info(
+                "write_input was called, but input has already been written."
```
Comment (jan-janssen) [inline pyiron_base/jobs/job/generic.py]:
```suggestion
                "write_input was called, but working_directory is not empty."
```

Comment (Leimeroth) [reply pyiron_base/jobs/job/generic.py]: added a check to the write_input_from_input_dict function

**Comment by Leimeroth** [comment]
Ok, so check if working directory is empty leads to a bunch of test failures and some horrible side effects, because I created the job with delete_existing_job=True. However the contents of the working directory are not deleted on time,
so on my local computer no input is written with the check.

**Comment by jan-janssen** [comment]
> Ok, so check if working directory is empty leads to a bunch of test failures and some horrible side effects, because I created the job with delete_existing_job=True. However the contents of the working directory are not deleted on time, so on my local computer no input is written with the check.

The `!=` and `==` sign were mixed up. You want to write the input files when the directory is empty and do not write them when it is not empty.

**Comment by jan-janssen** [comment]
@Leimeroth Can you run your tests again, like submitting a job with a custom potential to a remote cluster? Once this works and you are fine with the changes as they are in this pull request, can you approve this pull request? Then I can make sure it is included in the release of the next version. I aim to update both `pyiron_base` and `pyiron_atomistics` before the pyiron meeting on Monday.

**Comment by Leimeroth** [comment]
Now it works with my own potential and the EAM.
However, only the check prevents it from rewriting input, which could cause trouble when trying to rerun jobs.
When I run them again using delete_existing_job=True on my local computer the old files are still present on the remote cluster, so they will be used instead of my updated inputs.

**Comment by jan-janssen** [comment]
Are you sure about this? I thought the files are written locally and then transferred. So as the job is first deleted locally before the new input files are written the working directory is removed and a new empty working directory is created. Then at the copying stage we do not use the `write_input()` function so it should still overwrite the files on the cluster.

**Comment by Leimeroth** [comment]
> Are you sure about this? I thought the files are written locally and then transferred. So as the job is first deleted locally before the new input files are written the working directory is removed and a new empty working directory is created. Then at the copying stage we do not use the `write_input()` function so it should still overwrite the files on the cluster.

You are right, I just tested it. Fine to merge from my side.

**Pull request merged**